File: guides/micronaut-security-jwt/java/src/main/java/example/micronaut/CustomRefreshTokenPersistence.java
package example.micronaut;

import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.errors.OauthErrorResponseException;
import io.micronaut.security.token.event.RefreshTokenGeneratedEvent;
import io.micronaut.security.token.refresh.RefreshTokenPersistence;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;

import java.util.Optional;

import static io.micronaut.security.errors.IssuingAnAccessTokenErrorCode.INVALID_GRANT;

@Singleton // <1>
public class CustomRefreshTokenPersistence implements RefreshTokenPersistence {

    private final RefreshTokenRepository refreshTokenRepository;

    public CustomRefreshTokenPersistence(RefreshTokenRepository refreshTokenRepository) { // <2>
        this.refreshTokenRepository = refreshTokenRepository;
    }

    @Override
    public void persistToken(RefreshTokenGeneratedEvent event) { // <3>
        if (event != null &&
                event.getRefreshToken() != null &&
                event.getAuthentication() != null &&
                event.getAuthentication().getName() != null) {
            String payload = event.getRefreshToken();
            refreshTokenRepository.save(event.getAuthentication().getName(), payload, false); // <4>
        }
    }

    @Override
    public Publisher<Authentication> getAuthentication(String refreshToken) {
        return Flux.create(emitter -> {
            Optional<RefreshTokenEntity> tokenOpt = refreshTokenRepository.findByRefreshToken(refreshToken);
            if (tokenOpt.isPresent()) {
                RefreshTokenEntity token = tokenOpt.get();
                if (token.getRevoked()) {
                    emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token revoked", null)); // <5>
                } else {
                    emitter.next(Authentication.build(token.getUsername())); // <6>
                    emitter.complete();
                }
            } else {
                emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token not found", null)); // <7>
            }
        }, FluxSink.OverflowStrategy.ERROR);
    }
}

File: guides/micronaut-security-jwt/python/src/example/micronaut/custom_refresh_token_persistence.py
from jakarta.inject import Singleton
from micronaut.security.authentication import Authentication
from micronaut.security.errors import (
    IssuingAnAccessTokenErrorCode,
    OauthErrorResponseException,
)
from micronaut.security.token.refresh import RefreshTokenPersistence
from org.reactivestreams import Publisher
from reactor.core.publisher import Flux, FluxSink

from .refresh_token_repository import RefreshTokenRepository


# tag::clazz[]
@Singleton  # <1>
class CustomRefreshTokenPersistence(RefreshTokenPersistence):
    def __init__(self, refresh_token_repository: RefreshTokenRepository):  # <2>
        self.refresh_token_repository = refresh_token_repository

    def persistToken(self, event) -> None:  # <3>
        if (
            event is not None
            and event.getRefreshToken() is not None
            and event.getAuthentication() is not None
            and event.getAuthentication().getName() is not None
        ):
            payload = event.getRefreshToken()
            self.refresh_token_repository.save(event.getAuthentication().getName(), payload, False)  # <4>

    def getAuthentication(self, refreshToken: str) -> Publisher:
        def emit(emitter):
            token_optional = self.refresh_token_repository.findByRefreshToken(refreshToken)
            if token_optional.isPresent():
                token = token_optional.get()
                if token.revoked:
                    emitter.error(
                        OauthErrorResponseException(
                            IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                            "refresh token revoked",
                            None,
                        )
                    )  # <5>
                else:
                    emitter.next(Authentication.build(token.username))  # <6>
                    emitter.complete()
            else:
                emitter.error(
                    OauthErrorResponseException(
                        IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                        "refresh token not found",
                        None,
                    )
                )  # <7>

        return Flux.create(emit, FluxSink.OverflowStrategy.ERROR)
# end::clazz[]
